packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,246 @@
-import { sendRequest, HttpMethod } from '@biconomy/common'
+import { Logger, sendRequest, HttpMethod, getTimestampInSeconds } from '@biconomy/common'
 import { resolveProperties } from '@ethersproject/properties'
-import { UserOperation } from '@biconomy/core-types'
-import { IPaymaster } from 'interfaces/IPaymaster'
-import { PaymasterAndDataResponse, PaymasterConfig } from './utils/Types'
+import { UserOperation, Transaction } from '@biconomy/core-types'
+import { Provider } from '@ethersproject/abstract-provider'
+import {
+  PaymasterFeeQuote,
+  PaymasterConfig,
+  BiconomyTokenPaymasterFeeQuoteResponse,
+  FeeQuotesOrDataDto,
+  SponsorUserOperationDto,
+  PaymasterServiceSuccessResponse,
+  BiconomyTokenPaymasterRequest,
+  PaymasterMode
+} from './utils/Types'
+import { BigNumberish, BigNumber, ethers } from 'ethers'
+import { ERC20_ABI } from './constants'
+import { IHybridPaymaster } from './interfaces/IHybridPaymaster'
 
-export class BiconomyPaymaster implements IPaymaster {
+// Hybrid - Generic Gas Abstraction paymaster
+/**
+ *
+ */
+export class BiconomyPaymaster implements IHybridPaymaster {
   constructor(readonly paymasterConfig: PaymasterConfig) {}
+
+  // Note: maybe rename to createTokenApprovalTransaction
+  async createTokenApprovalRequest(
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest,
+    provider: Provider
+  ): Promise<Transaction> {
+    const feeTokenAddress: string = tokenPaymasterRequest.feeQuote.tokenAddress
+    Logger.log('erc20 fee token address ', feeTokenAddress)
+
+    const spender = tokenPaymasterRequest.spender
+    Logger.log('spender address ', spender)
+
+    // TODO
+    // Note: should also check in caller if the approval is already given, if yes return object with address or data 0
```

Comment:
That's fine approval check can be provided as separate method. A function should do one thing only and return consistent response.

Reply:
ok, will write different helper to read